src/bucket/Bucket.h

```diff
@@ -34,29 +35,66 @@ class Database;
 class Bucket : public std::enable_shared_from_this<Bucket>,
                public NonMovableOrCopyable
 {
-
-    std::string const mFilename;
+    std::filesystem::path const mFilename;
     Hash const mHash;
     size_t mSize{0};
+    bool mLazyIndex;
+
+    std::unique_ptr<BucketIndex const> mIndex{};
+
+    // Lazily-constructed and retained for read path.
+    std::unique_ptr<XDRInputFileStream> mStream;
 
   public:
     // Create an empty bucket. The empty bucket has hash '000000...' and its
     // filename is the empty string.
     Bucket();
 
+    // Returns true if bucket is indexed, false otherwise
+    bool isIndexed() const;
```

Comment:
Currently, this is only called in a few places where a class needs to decide whether or not to index a bucket when indexing the entire bucketlist synchronously. I don't know if there is a good way to do this without this method. In several startup paths, we need to index the entire bucket list twice. For example, if we are running a catchup but already have some buckets saved on disk, we need to index the bucketlist initially to assume LCL state, then again after the bucket apply step. If the catchup is over a small range, many of the bucket indexes from the LCL will still be valid, so we use `isIndexed` to avoid indexing redundantly. We could adjust the abstraction some and replace this with `indexIfnotIndexed` method, but I think we'll need some way of outside classes knowing whether or not a bucket is indexed for these scenarios.